Indirect solvers

nrontsis opened the pull request:
Adds support for iterative KKT solvers like cg and minres for both the reduced and the full KKT linear system.

Required dependecies: `IterativeSolvers.jl`, `LinearMaps.jl`

migarstka commented on src/kktsolver_indirect.jl:
```diff
@@ -0,0 +1,170 @@
+    # And then recover y2 as
+    # y2 = ρ(Ay1 - x2)
+
+    x1 = view(x, 1:S.n); y1 = view(y, 1:S.n)
```
The input vectors `x` and `y` (named `sol` and `lhs` in `admm_step!()` in `solver.jl`) are always the same size and occupy the same location in memory. Therefore, I would suggest to create the views `x1`, `x2`, `y1` and `y2` when the solver object is created and store them inside the object. This avoids recreating the views at every `solve!()`.

nrontsis replied:
Sure

migarstka commented on src/kktsolver_indirect.jl:
```diff
@@ -0,0 +1,170 @@
+    y1 .+= x1
+
+    push!(S.multiplications, 0)
+    function reduced_mul!(y::AbstractVector{T}, x::AbstractVector{T}) where {T}
```
Is this as efficient as "pulling" the function "out"?

nrontsis replied:
Since we use `IterativeSolvers.jl` without an explicit matrix, we have to rely on the `LinearMaps.jl` package. To create a `LinearMap`, one should provide a function of the form
```julia
reduced_mul!(y::AbstractVector, x::AbstractVector)
```
i.e. a function with *only* inputs the input/output vectors.

I could define a function
```julia
 function reduced_mul!(y::AbstractVector{T}, x::AbstractVector{T}, S::IndirectReducedKKTSolver) where {T}
```
outside of `solve!()`, but then I would need to create the `LinearMap` with another wrapper function i.e.
```julia
L = LinearMap((x, y) -> reduced_mul!(x, y, S), S.n; ismutating=true, issymmetric=true)
```

Thus, I don't see the benefits of pulling `reduced_mul!` out of `solve!` because we have to create a function at every call of `solve!`. Perhaps I am missing something?

migarstka commented:
Thanks Nikitas! It's great to have support for CG and MINRES. 

- Now that we support QDLDL, Cholmod, Pardiso, CG and MINRES I think we should probably make the documentation a bit more extensive about how to use them, etc.
- I will create a Jupyter notebook that shows how to use all of the different options the user has.

migarstka commented on src/kktsolver_indirect.jl:
```diff
@@ -0,0 +1,170 @@
+    return y
+end
+
+mutable struct IndirectKKTSolver{TP, TA, T} <: COSMO.AbstractKKTSolver
```
If the only option here is MINRES, maybe this should be named `MINRESIndirectKKTSolver` and automatically call minres without the extra `solver_type` option. What do you think?

migarstka replied:
Actually, another idea would be to create two new type aliases:


```julia
CGIndirectKKTSolver = IndirectReducedKKTSolver(..., solver_type = :CG, ...)
MINRESIndirectKKTSolver = IndirecKKTSolver(..., solver_type = :MINRES, ...)
```

nrontsis commented:
Thanks for the feedback! I have just pushed changes that implement your suggestions.

Continuous Integration does not run the tests in this PR  as the tests require `IterativeSolvers.jl` and `LinearMaps.jl` to run. What would be the best way to address this? I believe that this is also happening for `Pardiso`'s tests.

nrontsis commented:
Actually, my commit errors. I didn't see the error because I did not have `IterativeSolvers.jl` installed in my environment and thus the respective tests did not run :scream:

Apparently, [subarrays](https://docs.julialang.org/en/v1/devdocs/subarrays/#SubArray-design-1) (a.k.a. views) are immutable structs, thus their parent array cannot change.

@migarstka I am leaving my erroneous commit in case you have any other suggestions that might make it work. If not, I will revert it back to the way I had it before.

migarstka commented:
Thanks for the update Nikitas.

**Test dependencies**
I believe we can update `Project.toml` to  have test-specific dependencies:
https://julialang.github.io/Pkg.jl/latest/creating-packages/#Test-specific-dependencies-in-Julia-1.0-and-1.1-1

**Subarrays**
The following code works:
```julia

mutable struct Vect
  a::Vector{Int64}
  v::SubArray

  function Vect()
    a = collect(1:10)
    v = view(a, 1:5)
    new(a, v)
  end
end

test_vec = Vect()
println("a: $(test_vec.a), v: $(test_vec.v)")

# change a
test_vec.a[1] = -1
println("a: $(test_vec.a), v: $(test_vec.v)")

# change v
test_vec.v[1] = 100
println("a: $(test_vec.a), v: $(test_vec.v)")
```
Is what you want to do different to that?

nrontsis commented:
I was hoping to change the `test_vec` of your example as following: 
```julia
test_vec.a = Vector(-10:-1)
```
i.e. to point to a *new* vector. When doing this, however, `test_vec.v` does not view into `test_vec.a` any more.


I guess I could do something like
```julia
for i = 1:n
     test_vec.a[i] = new_vector[i]
end
```
i.e. copy the parts of the `new_vector` to the predefined `test_vec` instead of creating new views at every iteration (as I do now).

This would not allocate new memory at every iteration. However, I believe that it would be slower, according to the following benchmark script
```julia
using BenchmarkTools
function my_copy!(x1, x, n)
     @inbounds @simd for i = 1:n
           x1[i] = x[i]
     end
end
n = 10^7
x = randn(2*n)
x1 = randn(n)
@btime view(x, 1:n);
@btime my_copy!(x1, x, n);
```
which gives 36.598 ns for a view and 12.561 ms for a copy.

nrontsis commented:
Following our discussion, I reverted to the previous commit, enabled the tests in CI, and added the aliases you suggested.

Enabling Pardiso's tests is tracked in #99.

I believe that CI fails on some MOI test due to an out-of-date `Project.toml` list. I am not sure if this should be fixed here on another PR.
